File: ios/Sources/Omnesis/Permissions/MobileSourceActivation.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation

/// What a phone must establish before it asks the operating system for access
/// to a device-hosted source. This is source-agnostic: the descriptor's mode
/// decides whether two devices are redundant views, independent partitions,
/// or mutually exclusive hosts.
public enum MobileSourceActivationPlan: Equatable, Sendable {
    case ready
    case join
    case addPartition
    case choose(SourceMultiDeviceMode)
    case incompatible(current: SourceMultiDeviceMode, desired: SourceMultiDeviceMode)
}

public enum MobileSourceActivationChoice: Equatable, Sendable {
    case keepOther
    case useBoth
    case takeOver
}

public enum MobileSourceActivationOutcome: Equatable, Sendable {
    case ready
    case keptOther
    case choiceRequired(SourceMultiDeviceMode)
    case incompatible(current: SourceMultiDeviceMode, desired: SourceMultiDeviceMode)
}

/// How much of a source's data the operating system lets this device read.
public enum MobileSourceGrant: Equatable, Sendable {
    /// Everything the source reads, including while Omnesis is closed.
    case full
    /// Only a subset the user picked, such as selected photos.
    case limited
    /// Only while Omnesis is open, such as location While Using the App.
    case foregroundOnly
}

/// Where a source's operating-system permission request settled.
public enum MobileSourceAuthorization: Equatable, Sendable {
    case granted(MobileSourceGrant)
    case notAllowed
    case unavailable(reason: String)
    /// The request itself could not run; nothing was decided.
    case failed(message: String?)
}

/// A source's own work around the shared activation path: asking the
/// operating system for access, switching the source on locally, and anything
/// that has to wait for the rebuilt collector.
public struct MobileSourceActivationSteps {
    public let authorize: @MainActor () async -> MobileSourceAuthorization
    public let activate: @MainActor () async -> Void
    public let afterActivation: @MainActor () async -> Void

    public init(
        authorize: @escaping @MainActor () async -> MobileSourceAuthorization,
        activate: @escaping @MainActor () async -> Void,
        afterActivation: @escaping @MainActor () async -> Void = {}
    ) {
        self.authorize = authorize
        self.activate = activate
        self.afterActivation = afterActivation
    }
}

/// The result of turning on one phone-hosted source.
public enum MobileSourceEnableResult: Equatable, Sendable {
    case enabled(MobileSourceGrant)
    /// The operating system refused access, so nothing was changed.
    case notAllowed
    /// This device cannot provide the source.
    case unavailable(reason: String)
    /// The user chose to leave the source with the device that hosts it.
    case keptOther
    case choiceRequired(SourceMultiDeviceMode)
    /// Something else stopped the change; `message` is readable copy when known.
    case failed(message: String?)
}

public struct MobileSourceActivationOperations: Sendable {
    public let setMode: @Sendable (String, SourceMultiDeviceMode) async throws -> Void
    public let join: @Sendable (String, String) async throws -> Void
    public let transfer: @Sendable (String, String) async throws -> Void

    public init(
        setMode: @escaping @Sendable (String, SourceMultiDeviceMode) async throws -> Void,
        join: @escaping @Sendable (String, String) async throws -> Void,
        transfer: @escaping @Sendable (String, String) async throws -> Void
    ) {
        self.setMode = setMode
        self.join = join
        self.transfer = transfer
    }
}

public enum MobileSourceActivationError: LocalizedError, Equatable, Sendable {
    case invalidChoice(MobileSourceActivationChoice, SourceMultiDeviceMode)
    case notContributing
    case incompatibleMode(SourceMultiDeviceMode, SourceMultiDeviceMode)
    /// Another device changed who hosts the source while this one was
    /// activating it.
    case hostChanged

    public var errorDescription: String? {
        switch self {
        case .hostChanged:
            "Another device started sending this source. Try again to choose how this iPhone contributes."
        case .invalidChoice(let choice, let mode):
            "The \(choice) choice is not valid for a \(mode.rawValue) source."
        case .notContributing:
            "This device is not contributing to this source. Enable it before syncing."
        case .incompatibleMode(let current, let desired):
            "This source uses \(current.rawValue) mode; this device requires \(desired.rawValue) mode."
        }
    }
}

public enum MobileSourceActivation {
    public static func plan(
        source: SourceRecord?,
        deviceId: String,
        desiredMode: SourceMultiDeviceMode
    )
        -> MobileSourceActivationPlan {
        guard let source else { return .ready }
        let currentMode = SourceMultiDeviceMode(rawValue: source.multiDeviceMode ?? "exclusive")
            ?? .exclusive
        if currentMode == .exclusive, desiredMode == .partitioned { return .addPartition }
        if desiredMode == .partitioned, currentMode != desiredMode {
            return .incompatible(current: currentMode, desired: desiredMode)
        }
        if source.hosts(deviceId) { return .ready }
        if currentMode == desiredMode {
            return currentMode == .exclusive ? .choose(.exclusive) : .join
        }
        if currentMode == .exclusive {
            return desiredMode == .partitioned ? .addPartition : .choose(desiredMode)
        }
        return .incompatible(current: currentMode, desired: desiredMode)
    }

    /// Prepare an already-enabled independent stream before reading its
    /// cursor. Only an existing host may adopt legacy shared history here;
    /// ordinary background sync never rejoins a detached phone or unpauses a
    /// source. The explicit mode request must finish before any data is read.
    public static func prepareHostedPartition(
        source: SourceRecord?,
        deviceId: String,
        setMode: @Sendable (String, SourceMultiDeviceMode) async throws -> Void
    ) async throws {
        guard let source, source.hosts(deviceId) else {
            throw MobileSourceActivationError.notContributing
        }
        let currentMode = SourceMultiDeviceMode(rawValue: source.multiDeviceMode ?? "exclusive") ?? .exclusive
        switch currentMode {
        case .partitioned: return
        case .exclusive: try await setMode(source.id, .partitioned)
        case .replicated: throw MobileSourceActivationError.incompatibleMode(currentMode, .partitioned)
        }
    }

    /// Only explicit activation may create a registration. A background
    /// refresh using a stale live row could otherwise clear a newer tombstone.
    public static func mayCreateRegistration(
        source: SourceRecord?,
        deviceId: String,
        sourceId: String? = nil,
        allowCreationFor: String? = nil
    )
        -> Bool {
        guard let requestedId = sourceId ?? source?.id, requestedId == allowCreationFor else { return false }
        return source.map { $0.hosts(deviceId) } ?? true
    }

    /// Resolve the user-visible decision without mutating gateway ownership.
    /// Protected-data authorization can safely happen after this returns
    /// `.ready`; callers commit with `execute` only after the OS grant.
    public static func inspect(
        source: SourceRecord?,
        deviceId: String,
        desiredMode: SourceMultiDeviceMode,
        choice: MobileSourceActivationChoice? = nil
    ) throws
        -> MobileSourceActivationOutcome {
        let activationPlan = plan(source: source, deviceId: deviceId, desiredMode: desiredMode)
        switch activationPlan {
        case .ready, .join, .addPartition:
            return .ready
        case .incompatible(let current, let desired):
            return .incompatible(current: current, desired: desired)
        case .choose(let currentMode):
            guard let choice else { return .choiceRequired(currentMode) }
            switch choice {
            case .keepOther:
                return .keptOther
            case .useBoth:
                guard desiredMode != .exclusive, currentMode == desiredMode else {
                    throw MobileSourceActivationError.invalidChoice(choice, currentMode)
                }
                return .ready
            case .takeOver:
                let persistedMode = SourceMultiDeviceMode(
                    rawValue: source?.multiDeviceMode ?? "exclusive"
                ) ?? .exclusive
                guard persistedMode == .exclusive else {
                    throw MobileSourceActivationError.invalidChoice(choice, persistedMode)
                }
                return .ready
            }
        }
    }

    public static func execute(
        source: SourceRecord?,
        deviceId: String,
        desiredMode: SourceMultiDeviceMode,
        choice: MobileSourceActivationChoice? = nil,
        operations: MobileSourceActivationOperations
    ) async throws
        -> MobileSourceActivationOutcome {
        let activationPlan = plan(source: source, deviceId: deviceId, desiredMode: desiredMode)
        guard let source else { return .ready }

        switch activationPlan {
        case .ready:
            return .ready
        case .join:
            try await operations.join(source.id, deviceId)
            return .ready
        case .addPartition:
            try await operations.setMode(source.id, .partitioned)
            try await operations.join(source.id, deviceId)
            return .ready
        case .incompatible(let current, let desired):
            return .incompatible(current: current, desired: desired)
        case .choose(let currentMode):
            guard let choice else { return .choiceRequired(currentMode) }
            switch choice {
            case .keepOther:
                return .keptOther
            case .useBoth:
                guard desiredMode != .exclusive, currentMode == desiredMode else {
                    throw MobileSourceActivationError.invalidChoice(choice, currentMode)
                }
                try await operations.setMode(source.id, desiredMode)
                try await operations.join(source.id, deviceId)
                return .ready
            case .takeOver:
                let persistedMode = SourceMultiDeviceMode(rawValue: source.multiDeviceMode ?? "exclusive")
                    ?? .exclusive
                guard persistedMode == .exclusive else {
                    throw MobileSourceActivationError.invalidChoice(choice, persistedMode)
                }
                try await operations.transfer(source.id, deviceId)
                return .ready
            }
        }
    }
}

File: android/core-transport/src/main/kotlin/dev/omnesis/android/transport/MobileSourceActivation.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.transport

import dev.omnesis.android.transport.dto.SourceRecord

sealed interface ActivationPlan {
    data object Ready : ActivationPlan
    data object Join : ActivationPlan
    data object AddPartition : ActivationPlan
    data class Choose(val mode: SourceMultiDeviceMode) : ActivationPlan
    data class Incompatible(
        val current: SourceMultiDeviceMode,
        val desired: SourceMultiDeviceMode,
    ) : ActivationPlan
}

enum class ActivationChoice { KEEP_OTHER, USE_BOTH, TAKE_OVER }

sealed interface ActivationOutcome {
    data object Ready : ActivationOutcome
    data object KeptOther : ActivationOutcome
    data class ChoiceRequired(val mode: SourceMultiDeviceMode) : ActivationOutcome
    data class Incompatible(
        val current: SourceMultiDeviceMode,
        val desired: SourceMultiDeviceMode,
    ) : ActivationOutcome
}

data class ActivationOperations(
    val setMode: suspend (String, SourceMultiDeviceMode) -> Unit,
    val join: suspend (String, String) -> Unit,
    val transfer: suspend (String, String) -> Unit,
)

class InvalidActivationChoice(choice: ActivationChoice, mode: SourceMultiDeviceMode) :
    IllegalArgumentException("The $choice choice is not valid for a ${mode.wireValue} source")

class SourceModeTransitionNotConfirmed(mode: SourceMultiDeviceMode) :
    IllegalStateException("The gateway did not confirm ${mode.wireValue} mode. Update the gateway, then try again.")

/**
 * Turns a source-contract refusal into the gateway's actionable guidance while
 * leaving connectivity, authentication, and unexpected failures to the
 * source screen's context-specific fallback. The gateway message identifies
 * the incompatible device; preserving it is more useful than misreporting a
 * deliberate 409 as a network failure.
 */
fun activationFailureMessage(failure: Throwable, fallback: String): String {
    if (failure is SourceModeTransitionNotConfirmed) return failure.message ?: fallback
    val serverError = failure as? GatewayException.ServerError ?: return fallback
    if (serverError.code != "MULTI_DEVICE_CONTRACT_UNSUPPORTED") return fallback
    return serverError.body
        ?.trim()
        ?.takeIf(String::isNotEmpty)
        ?.replaceFirstChar { it.uppercase() }
        ?: "Another device hosting this source must update Omnesis before this phone can join it."
}

/**
 * One gateway half of a phone source's explicit enable, reduced to what the
 * screen that asked acts on: carry on, stop because the user kept another
 * device, ask the user how this phone should contribute, or say why it failed.
 */
sealed interface ActivationStep {
    data object Ready : ActivationStep
    data object KeptOther : ActivationStep
    data class ChoiceRequired(val mode: SourceMultiDeviceMode) : ActivationStep
    data class Failed(val message: String) : ActivationStep
}

const val ACTIVATION_PREPARE_FAILED = "The gateway could not prepare this phone. Check the connection and try again."
const val ACTIVATION_COMMIT_FAILED = "The gateway could not finish enabling this phone. Try again."
const val ACTIVATION_INCOMPATIBLE = "This source uses a different multi-device mode on the gateway."

/** The inspection before Android is asked for access, with every failure folded into a step. */
suspend fun SourceMembership.inspectActivationStep(
    sourceId: String,
    desiredMode: SourceMultiDeviceMode,
    choice: ActivationChoice? = null,
): ActivationStep = try {
    when (val outcome = inspectActivation(sourceId, desiredMode, choice)) {
        ActivationOutcome.Ready -> ActivationStep.Ready
        ActivationOutcome.KeptOther -> ActivationStep.KeptOther
        is ActivationOutcome.ChoiceRequired -> ActivationStep.ChoiceRequired(outcome.mode)
        is ActivationOutcome.Incompatible -> ActivationStep.Failed(ACTIVATION_INCOMPATIBLE)
    }
} catch (e: kotlin.coroutines.cancellation.CancellationException) {
    throw e
} catch (e: Exception) {
    ActivationStep.Failed(activationFailureMessage(e, ACTIVATION_PREPARE_FAILED))
}

/**
 * The commit after Android granted access. A decision that changed while the
 * system dialog was open cannot be carried out on the user's earlier answer,
 * so it fails rather than asking again behind their back.
 */
suspend fun SourceMembership.commitActivationStep(
    sourceId: String,
    desiredMode: SourceMultiDeviceMode,
    choice: ActivationChoice? = null,
): ActivationStep = try {
    when (commitActivation(sourceId, desiredMode, choice)) {
        ActivationOutcome.Ready -> ActivationStep.Ready
        ActivationOutcome.KeptOther -> ActivationStep.KeptOther
        is ActivationOutcome.ChoiceRequired,
        is ActivationOutcome.Incompatible,
        -> ActivationStep.Failed(ACTIVATION_COMMIT_FAILED)
    }
} catch (e: kotlin.coroutines.cancellation.CancellationException) {
    throw e
} catch (e: Exception) {
    ActivationStep.Failed(activationFailureMessage(e, ACTIVATION_COMMIT_FAILED))
}

/** Determines the device-owned enable flow from the persisted source contract. */
object MobileSourceActivation {
    /**
     * Prepare an already-contributing device without joining it implicitly.
     * A missing source/member is not permission to recreate it in the background.
     */
    suspend fun prepareHostedPartition(
        source: SourceRecord?,
        deviceId: String,
        setMode: suspend (String, SourceMultiDeviceMode) -> Unit,
    ) {
        check(source != null && source.hosts(deviceId)) { "This device is not contributing to the source" }
        when (source.multiDeviceMode ?: SourceMultiDeviceMode.EXCLUSIVE.wireValue) {
            SourceMultiDeviceMode.PARTITIONED.wireValue -> Unit
            SourceMultiDeviceMode.EXCLUSIVE.wireValue -> setMode(source.id, SourceMultiDeviceMode.PARTITIONED)
            else -> throw IllegalStateException("This source requires a device-local partition before syncing")
        }
    }

    fun plan(
        source: SourceRecord?,
        deviceId: String,
        desiredMode: SourceMultiDeviceMode,
    ): ActivationPlan {
        if (source == null) return ActivationPlan.Ready
        val currentMode = SourceMultiDeviceMode.entries.firstOrNull {
            it.wireValue == (source.multiDeviceMode ?: SourceMultiDeviceMode.EXCLUSIVE.wireValue)
        } ?: SourceMultiDeviceMode.EXCLUSIVE
        // A device-local source must use its own storage stream even when
        // this device already owns a legacy exclusive source. Do not apply
        // this upgrade rule to shared sources whose owner chose exclusivity.
        if (desiredMode == SourceMultiDeviceMode.PARTITIONED) {
            if (currentMode == SourceMultiDeviceMode.EXCLUSIVE) return ActivationPlan.AddPartition
            if (currentMode != desiredMode) return ActivationPlan.Incompatible(currentMode, desiredMode)
        }
        if (source.hosts(deviceId)) return ActivationPlan.Ready
        if (currentMode == desiredMode) {
            return if (currentMode == SourceMultiDeviceMode.EXCLUSIVE) {
                ActivationPlan.Choose(SourceMultiDeviceMode.EXCLUSIVE)
            } else {
                ActivationPlan.Join
            }
        }
        if (currentMode == SourceMultiDeviceMode.EXCLUSIVE) {
            return if (desiredMode == SourceMultiDeviceMode.PARTITIONED) {
                ActivationPlan.AddPartition
            } else {
                ActivationPlan.Choose(desiredMode)
            }
        }
        return ActivationPlan.Incompatible(currentMode, desiredMode)
    }

    suspend fun execute(
        source: SourceRecord?,
        deviceId: String,
        desiredMode: SourceMultiDeviceMode,
        choice: ActivationChoice? = null,
        operations: ActivationOperations,
    ): ActivationOutcome {
        val activationPlan = plan(source, deviceId, desiredMode)
        if (source == null) return ActivationOutcome.Ready
        return when (activationPlan) {
            ActivationPlan.Ready -> ActivationOutcome.Ready
            ActivationPlan.Join -> {
                operations.join(source.id, deviceId)
                ActivationOutcome.Ready
            }
            ActivationPlan.AddPartition -> {
                operations.setMode(source.id, SourceMultiDeviceMode.PARTITIONED)
                operations.join(source.id, deviceId)
                ActivationOutcome.Ready
            }
            is ActivationPlan.Incompatible -> ActivationOutcome.Incompatible(
                activationPlan.current,
                activationPlan.desired,
            )
            is ActivationPlan.Choose -> when (choice) {
                null -> ActivationOutcome.ChoiceRequired(activationPlan.mode)
                ActivationChoice.KEEP_OTHER -> ActivationOutcome.KeptOther
                ActivationChoice.USE_BOTH -> {
                    if (desiredMode == SourceMultiDeviceMode.EXCLUSIVE || activationPlan.mode != desiredMode) {
                        throw InvalidActivationChoice(choice, activationPlan.mode)
                    }
                    operations.setMode(source.id, desiredMode)
                    operations.join(source.id, deviceId)
                    ActivationOutcome.Ready
                }
                ActivationChoice.TAKE_OVER -> {
                    val persistedMode = SourceMultiDeviceMode.entries.firstOrNull {
                        it.wireValue == (source.multiDeviceMode ?: SourceMultiDeviceMode.EXCLUSIVE.wireValue)
                    } ?: SourceMultiDeviceMode.EXCLUSIVE
                    if (persistedMode != SourceMultiDeviceMode.EXCLUSIVE) {
                        throw InvalidActivationChoice(choice, persistedMode)
                    }
                    operations.transfer(source.id, deviceId)
                    ActivationOutcome.Ready
                }
            }
        }
    }
}
